Open registration for c10 thread pool

jamesr66a opened the pull request:
1. Move ATen threadpool & open registration mechanism to C10
2. Move the `global_work_queue` to use this open registration mechanism, to allow users to substitute in their own

zdevito reviewed (approved):
OSS changes look fine.

dzhulgakov commented on c10/core/thread_pool.h:
```diff
@@ -100,8 +102,29 @@ class CAFFE2_API ThreadPool : public c10::TaskThreadPoolBase {
+      int numa_node_id = -1)
+      : ThreadPool(pool_size, numa_node_id) {}
+
+  // TODO move this to ATen/core/thread_pool.h
```
why?

jamesr66a replied:
I thought I deleted this comment

dzhulgakov commented on c10/core/thread_pool.h:
```diff
@@ -100,8 +102,29 @@ class CAFFE2_API ThreadPool : public c10::TaskThreadPoolBase {
+      : ThreadPool(pool_size, numa_node_id) {}
+
+  // TODO move this to ATen/core/thread_pool.h
+  void init_thread() override {
```
probably better move to .cpp file (and some includes too)

dzhulgakov commented on c10/core/thread_pool.cpp:
```diff
@@ -125,9 +124,32 @@ void setNumThreads(size_t v) {
+std::shared_ptr<TaskThreadPoolBase> createC10ThreadPool(
+    int device_id,
+    int pool_size,
+    bool create_new) {
```
enforce that create_new is false (as that's the only one supported)

Commit:
Open registration for c10 thread pool (#17788)

Summary:
1. Move ATen threadpool & open registration mechanism to C10
2. Move the `global_work_queue` to use this open registration mechanism, to allow users to substitute in their own
Pull Request resolved: https://github.com/pytorch/pytorch/pull/17788

Differential Revision: D14379707

fbshipit-source-id: 0bb1ce9560157e345647de46a632a15b42738308
